pdal/private/StageRunner.cpp

```diff
@@ -55,7 +55,22 @@ PointViewPtr StageRunner::keeps()
 void StageRunner::run()
 {
     point_count_t keepSize = m_keeps->size();
-    m_viewSet = m_stage->run(m_keeps);
+
+    // Don't run the stage if we're a filter and we're empty
+    bool bDoRunStage(true);
+    Filter* isFilterType = dynamic_cast<Filter*>(m_stage);
```

Comment:
@hobu I like that this logic is now in StageRunner, but I wonder if it makes sense for StageRunner to stay agnostic to the specific type of stage. Right now it’s checking dynamic_cast<Filter*>, which couples StageRunner to filters specifically.

Maybe instead we add a virtual method like skipIfEmpty() on Stage (default true), and have Filter override it if needed. Then StageRunner just calls that method and doesn’t need to know whether it’s dealing with a filter or something else. That way the skip logic stays centralized but type-agnostic.

Reply:
I don't think there are use cases right now for running filters if there are no points, so adding a virtual call that does exactly what this does doesn't add anything. This is simple and can be changed later if needs change.

Reply:
> Maybe instead we add a virtual method like skipIfEmpty() on Stage (default true), and have Filter override it if needed. 

I was trying to avoid adding another method because doing so would necessitate a major release. Waiting for feedback from @abellgithub about the entire approach, however. I know we've gone back and forth about this topic over the years.